### ISSUE TITLE
What's the purpose of the solved exercises in this repo?

### ISSUE THREAD

**Issue by asterite** [exercism/xcrystal#32]
Since users are to solve the exercises and contrast them with other users, when do they get to see the solutions in this repo?

And, what should the solutions in this repo show? Should they aim for clarity, brevity, performance?

I personally think, given that this is a website for learning to program, to use code that's as clear as possible, even if it's a bit longer, and efficient if possible.

(Yesterday I also read, just by chance, the [sample of 99 bottles of OOP](http://www.sandimetz.com/99bottles/sample) and I think the overall idea of the book should be applied to solutions here)

For example, I'll take a random exercise, [brackets push](https://github.com/exercism/xcrystal/blob/master/exercises/bracket-push/example.cr). Nothing against that code or the author, I'm just using it as an example to discuss this issue. The current code is this:

```cr
module Brackets
  extend self
  MAP = {'}' => '{', ']' => '[', ')' => '('}

  def are_valid?(str : String)
    create_stack(filter_brackets(str)).size == 0
  end

  private def filter_brackets(brackets : String) : Array(Char)
    brackets.chars.select { |c| (MAP.keys + MAP.values).includes?(c) }
  end

  private def create_stack(brackets : Array(Char)) : Array(Char)
    brackets.reduce([] of Char) do |stack, b|
      MAP[b]? && MAP[b]? == stack.last? ? stack[0..-2] : stack << b
    end
  end
end
```

One solution I came to is this:

```cr
module Brackets
  def self.valid?(string)
    stack = [] of Char
    string.each_char do |char|
      case char
      when '(', '[', '{'
        stack.push char
      when ')'
        return false unless stack.pop? == '('
      when ']'
        return false unless stack.pop? == '['
      when '}'
        return false unless stack.pop? == '{'
      end
    end
    stack.empty?
  end
end
```

It has the same number of lines, but:

* The algorithm is clear: we use a stack, push to it on open brackets, pop and check on closing brackets (we ignore other chars)
* There's just one method to follow, not three
* There are no type annotations: type annotations are OK for method overloads or sometimes for public APIs, but they are not mandatory and in most cases they make code longer and harder to read. For example, in `def valid?(string)` it's pretty clear that the argument should be a `String`, and that it will return a (kind-of) boolean value.
* It's efficient: the first algorithm uses a Hash (`MAP`), creates many intermediate structures (the call to `chars`, the call to `select`, `MAP.keys` and `MAP.values` create intermediate arrays, which are then added to create another intermediate array (and this is done for every character in the string!), etc.). If you benchmark both methods, the second version is 80 times faster, and allocates just one array for the stack of brackets.

So, my worry is that people take this code as idiomatic Crystal code. Later they find that it's maybe slow or hard to understand, and maybe that's not very good for the language.

**Comment by kotp** [comment]
**TL;DR**;  The code in the `example.cr` is "proof of concept" rather than idiomatic, but there is no reason it couldn't be idiomatic code.  It is not "front facing", not presented to the users of Exercism, but of course is available for those that are curious.

* * * *

Kytrina has mentioned that the example code in the repository has been "Proof Of Concept", and to proof the tests that are presented to the users of Exercism.

She has also mentioned that it may be beneficial to be idiomatic, which is definitely possible over time.

It is not code that is gathered by the client, so it is not "front facing" code, though, of course, it is available for anyone to view if/when they come to the repository.  I don't believe there is anything currently directing them to the example code at all, in the flow of the client and process of participating on the 'user' level.

It has been discussed that the tests should be as idiomatic as possible, as that is forward facing.

**Comment by asterite** [comment]
@kotp Thank you! I'll close this issue then. The many styles of programming will be seen when users solve the problems (I'll try to slowly solve them too) :-)

**Comment by mhelmetag** [comment]
@kotp , I would agree that the examples should probably be the best they can be (like a gold standard) but... that will happen over time.

@asterite , your solution was awesome to see! Thanks for that!

**Comment by elorest** [comment]
@asterite I originally wrote this as a single method with `if elsif` but then changed it to be more dry. I had gotten feedback on previous solutions that they should have a lot more methods "like the ruby one", so I also broke it up into more methods. I agree that it probably wasn't necessary though. 

One advantage to your solution is that it returns as soon as it's false and stops evaluating. :)
